fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/MutableRecordStoreState.java

```diff
@@ -122,13 +122,34 @@ public IndexState setState(@Nonnull String indexName, @Nonnull IndexState state)
         if ((users.get() & WRITE_MASK) == 0) {
             throw new RecordCoreException("record store state is not enabled for modification");
         }
-        IndexState previous;
+        IndexMetaData previous;
         if (state == IndexState.READABLE) {
-            previous = indexStateMap.get().remove(indexName);
+            previous = indexMetaDataMap.get().remove(indexName);
         } else {
-            previous = indexStateMap.get().put(indexName, state);
+            previous = indexMetaDataMap.get().put(indexName, new IndexMetaData(state));
         }
-        return previous == null ? IndexState.READABLE : previous;
+        return previous == null ? IndexState.READABLE : previous.getState();
+    }
+
+    /**
+     * Modify the meta-data of an index in this map.
+     * Caller should also modify the database to match.
+     * @param indexName the index name
+     * @param indexMetaData the new meta-data for the given index
+     * @return the previous the old meta-data of the given index
+     */
+    @Nullable
+    public IndexMetaData setIndexMetaData(@Nonnull String indexName, @Nonnull IndexMetaData indexMetaData) {
+        if ((users.get() & WRITE_MASK) == 0) {
+            throw new RecordCoreException("record store state is not enabled for modification");
+        }
+        IndexMetaData previous;
+        if (indexMetaData.getState() == IndexState.READABLE) {
+            previous = indexMetaDataMap.get().remove(indexName);
```

Comment:
I think the metadata should remain for `READABLE` indexes.  For example, it would be useful to know when it became readable.